arch/arm/core/aarch64/cache.S

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright 2020 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/*
+ * ARM64 Cortex-A Cache management
+ */
+
+#include <toolchain.h>
+#include <linker/sections.h>
+#include <arch/cpu.h>
+
+_ASM_FILE_PROLOGUE
+
+.macro	dcache_range_op, dcache op
+	/* Calculate dcache line size */
+	mrs	x3, ctr_el0
+	mov	x2, #4
+	ubfm	x3, x3, #16, #19
+	lsl	x2, x2, x3
+
+	/*
+	 * x2 = cacheline_size
+	 * x1 = start + end
+	 * x3 = cacheline_size - 1
+	 * x0 = x0 & ~(x3)
+	 */
+	add	x1, x0, x1
+	sub	x3, x2, #1
+	bic	x0, x0, x3
+1:
+	\dcache \op, x0
+	add	x0, x0, x2
+	cmp	x0, x1
+	blo	1b
+	dsb	sy
+.endm
+
+GTEXT(arch_dcache_flush)
+SECTION_FUNC(TEXT, arch_dcache_flush)
+	dcache_range_op dc civac
+	ret
+
+GTEXT(arch_dcache_invd)
+SECTION_FUNC(TEXT, arch_dcache_invd)
+	dcache_range_op dc ivac
+	ret
```

Comment:
In reality the v8 cores are not differentiating `ivac  `    `civac `  hope we do not have assumption anywhere in code ie: we must be tolerant to a writeback of dirty lines even when we are just intending to do  `ivac`.

Reply:
When executed at EL1, a DC IVAC instruction performs a clean and invalidate, meaning it performs the same
maintenance as a DC CIVAC instruction, if all of the following apply:
• EL2 is implemented and enabled in the current Security state.
• The value of HCR_EL2.VM is 1, meaning EL1&0 stage two address translation is enabled.

So I would keep ivac and civac seperate.

Reply:
Sure we can them separate. My intention was to highlight that who ever is using the `arch_dcache_invd ` should not assume there is no writeback.